src/vstest.console/CommandLine/Executor.cs

```diff
@@ -60,6 +61,20 @@ internal class Executor
     /// </summary>
     public Executor(IOutput output) : this(output, TestPlatformEventSource.Instance, new ProcessHelper(), new PlatformEnvironment())
     {
+        // TODO: Get rid of this by making vstest.console code properly async.
+        // The current implementation of vstest.console is blocking many threads that just wait
+        // for completion in non-async way. Because threadpool is setting the limit based on processor count,
+        // we exhaust the threadpool threads quickly when we set maxCpuCount to use as many workers as we have threads.
+        //
+        // This setting allow the threadpool to start start more threads than it normally would without any delay.
+        // This won't pre-start the threads, it just pushes the limit of how many are allowed to start without waiting,
+        // and in effect makes callbacks processed earlier, because we don't have to wait that much to receive the callback.
+        // The correct fix would be to re-visit all code that offloads work to threadpool and avoid blocking any thread,
+        // and also use async await when we need to await a completion of an action. But that is a far away goal, so this
+        // is a "temporary" measure to remove the threadpool contention.
+        var additionalThreadsCount = Environment.ProcessorCount * 4;
```

Comment:
I understand the concept of pushing the thread pool limit higher, but I'm not sure I get why we decided to raise the limit to exactly 4 times the processor count. Is "4" a constant determined empirically that yielded the best throughput overall or is there a more formal reason for choosing it ?

Reply:
So I ended up with 1* (workerThreads) + 4*, to land on 5*

Reply:
I would write that down as comment in code for later.

Reply:
Have you tried to chance only the worker one? I don't know if chance the I/O makes difference here, we're not async at all so I suppose we're not using it.

Reply:
I did try to change, but if the IO pool is not exhausted then this option has no impact, it simply won't get past the limit. In both cases we are just telling the threadpool thread to start more threads without delay when it needs to. We will eventually  stop blocking them, and thread pool can re-balance itself, we are only paying a small price say 1MB per thread, but we know we will cap out on the amount of threads that is x times the processor count. So if there is a server that has 100 cores, it probably will also have 1GB of ram to spare. If it ever gets to the point where it allocates that many threads.

Reply:
Added the comment.